[Canvas] Adds doc links and keyboard shortcut cheatsheat to help menu

cqliu1 opened the pull request:
Closes #31266.
Closes #29657.

This moves doc links from the workpad loader to the global nav. This also adds a keyboard shortcut cheatsheet. These shortcuts are dynamically generated from `.../lib/keymap.js` and displays the shortcuts corresponding to the client OS.

### Global Nav Help Menu
<img width="328" alt="screen shot 2019-02-15 at 5 16 19 pm" src="https://user-images.githubusercontent.com/1697105/52891150-449b2200-3146-11e9-824d-4dd66fab6d8f.png">

### Mac Shortcuts
![screen shot 2019-02-15 at 3 39 22 pm](https://user-images.githubusercontent.com/1697105/52890946-cb4eff80-3144-11e9-976c-4c542fb355e1.png)

### Windows Shortcuts (same shortcuts for Linux and any other OS)
<img width="478" alt="screen shot 2019-02-15 at 4 56 11 pm" src="https://user-images.githubusercontent.com/1697105/52890959-da35b200-3144-11e9-8d3b-11addedfa587.png">

I'll add more shortcuts available in the layout engine in a future PR.

elasticmachine commented:
Pinging @elastic/kibana-canvas

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1021/)

monfera commented on x-pack/plugins/canvas/public/components/workpad_header/workpad_header.js:
```diff
@@ -80,7 +80,7 @@ export const WorkpadHeader = ({
               <FullscreenControl>
                 {({ toggleFullscreen }) => (
-                  <EuiToolTip position="bottom" content="Toggle fullscreen mode">
+                  <EuiToolTip position="bottom" content="Enter fullscreen mode">
```
Nice!

monfera commented on x-pack/plugins/canvas/public/components/workpad_page/index.js:
```diff
@@ -217,6 +217,16 @@ export const WorkpadPage = compose(
+  withHandlers(({ commit }) => ({
+    groupElements: () =>
+      commit('actionEvent', {
+        event: 'group',
```
👍

monfera commented on x-pack/plugins/canvas/public/lib/keymap.js:
```diff
@@ -38,36 +38,81 @@ const getCtrlShortcuts = shortcuts => {
+    },
+    GROUP: {
+      osx: ['g'],
+      windows: ['g'],
```
does it need 'G' and 'U' too?

monfera commented on x-pack/plugins/canvas/public/lib/keymap.js:
```diff
@@ -38,36 +38,81 @@ const getCtrlShortcuts = shortcuts => {
+      osx: ['u'],
+      windows: ['u'],
+      linux: ['u'],
+      other: ['u'],
```
If all platforms have the same key, would a single line of `other` cover it as a fallback, or is there some other defaulting? I'm not saying enlisting it is not nice, eg. for code documentation purpose.

cqliu1 replied:
It looks like react-shortcuts handles both cases. Both lowercase and uppercase keys (i.e with caps lock on) trigger the shortcuts.

cqliu1 replied:
Oh I needed the objects to be a specific shape. A single line would've worked too, but I wanted the shortcut to with the shape 
```javascript
{
help: 'description',
osx: ['shortcut'],
windows: ['shortcut'],
linux: ['shortcut'],
other: ['shortcut'],
}
``` 
because I needed the shortcut to be an object that I could add a `help` attribute to that is compatible with what `react-shortcuts` expects as a valid keymap.

monfera commented on x-pack/plugins/canvas/public/components/workpad_page/event_handlers.js:
```diff
@@ -163,16 +139,6 @@ const handleKeyDown = (commit, e, isEditable) => {
-const handleKeyPress = (commit, e, isEditable) => {
-  const { key, target } = e;
-  const upcaseKey = key && key.toUpperCase();
-  if (isEditable && !isTextInput(target) && 'GU'.indexOf(upcaseKey) !== -1) {
```
I see `isTextInput` is disused, is it the case that the new handler already isn't active when the input target is a text input tag?

cqliu1 replied:
I tested typing in an advanced filter and it appears that the handler doesn't correctly handle the event target being an input control. I'll add the check to the key handler.

monfera commented on x-pack/plugins/canvas/public/components/workpad_page/event_handlers.js:
```diff
@@ -126,30 +126,6 @@ const handleMouseDown = (commit, e, isEditable) => {
-
 const modifierKey = key => ['KeyALT', 'KeyCONTROL'].indexOf(keyCode(key)) > -1;
 
 const handleKeyDown = (commit, e, isEditable) => {
```
I haven't cleaned up some code here because the expectation was that, we're going to add back the initially present layout keyboard shortcuts for moving/rotating/resizing elements, which were initially fully handled in aero. Since we decided that keystrokes should be none of aero's concern, I recently removed the processing code, and I think this `handleKeyDown` is no longer doing anything useful so could be removed. May be extra safe to search for `keyboardEvent` under `aeroelastic/`. A bit tangential to your PR but you're reworking a good bunch of key handling so removing this vestige may free you up for whatever this may have blocked (or just remove a few lines).

cqliu1 replied:
I checked, and there aren't any instances of `keyboardEvent` in the aeroelastic code. I'll go ahead and remove the key event handlers

monfera commented:
With so many nicely documented shortcuts by now, a context menu will be very useful too :-)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1251/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1333/)

alexfrancoeur commented:
Great work @cqliu1! Some feedback below.

The docs button text looks much larger in comparison to the rest of Canvas. Is this specific to EUI? Maybe @ryankeairns typography changes addressed this.

![image](https://user-images.githubusercontent.com/25181823/53038334-7c59e080-344a-11e9-94a5-253e2e9580cd.png)

I'm guessing so, but are the commands conditional based on OS?

I still see the docs link in the list view. Was the removal of that link meant to be part of this PR as well?

![image](https://user-images.githubusercontent.com/25181823/53038501-e96d7600-344a-11e9-9c03-85267fd35627.png)

As this feature was snuck in so fast, I'm not sure what the expectations are. Are we meant to continually link out to general Kibana docs. I'd imagine so, I think it's helpful. That being said, i the primary action always meant to be the in context documentation? Would love to hear @cchaos & @ryankeairns thoughts here. FWIW I think the screenshot below makes sense, just want to make sure we're consistent with other implementations.

![image](https://user-images.githubusercontent.com/25181823/53038618-305b6b80-344b-11e9-86a3-70bed5ea7e5e.png)

ryankeairns commented:
@alexfrancoeur @cqliu1 

- The font size in the support menu is set globally, outside of Canvas. We're opting to use a smaller font size in the Canvas "chrome", thus the difference.
- We should remove that 'Docs' link from the workpad loader.
- The way the support menu works is that the top section (including the 'Go to documentation' button) is baked in for everybody. Plugin developers determine what appears below that section which in this case is the Canvas specific text and two buttons. 

The original PR for this support menu demonstrated plain text links, however the component was designed to accept whatever content is supplied. In talking with Catherine, the global nature/placement of the support button may not feel like a logical place to find Canvas docs so we discussed emphasizing our links visually - thus the buttons. It's tough, we may be setting an undesirable precedent using buttons here, but then the links feel visually insufficient on their own. 

For that reason, I was ok with the buttons, but perhaps a better design solution is needed in the near future.

cchaos commented:
@ryankeairns , The top section of the help popover is actually optional (on by default). You can opt to remove it by setting `useDefaultContent = false`
 https://github.com/elastic/kibana/blob/8aada890e578fd1a8849491188bba9b74870b3cc/src/legacy/ui/public/chrome/directives/header_global_nav/components/header_help_menu.tsx#L51

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1573/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1584/)

cqliu1 commented:
> I'm guessing so, but are the commands conditional based on OS?

Yes, they are. I was considering adding a radio button group so you can switch between Mac/Windows in the cheatsheet.

cqliu1 commented:
> I still see the docs link in the list view. Was the removal of that link meant to be part of this PR as well?

Yes, I forgot to remove it. I've gone ahead and removed the link in 16813e0.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1615/)

cqliu1 commented:
I added one minor change. I changed the arrow keys to arrow symbols in the cheatsheet.
<img width="482" alt="screen shot 2019-02-20 at 12 32 04 pm" src="https://user-images.githubusercontent.com/1697105/53119146-01f19500-350c-11e9-9202-6917618ca4a2.png">

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1625/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1630/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1848/)
